src/full_space.jl

```diff
@@ -0,0 +1,224 @@
+"""
+    FullSpaceModel{T, S} <: AbstractNLPModel{T, S}   
+
+A data type to store the full space `BlockNLPModel` as an `AbstractNLPModel`.
+"""
+mutable struct FullSpaceModel{T, S} <: AbstractNLPModel{T, S}
+    meta::NLPModelMeta{T, S}
+    counters::Counters
+    blocknlp::AbstractBlockNLPModel
+  end
+
+"""
+FullSpaceModel(
+    ::Type{T}, m::AbstractBlockNLPModel
+    ) where {T}
+
+Converts a `BlockNLPModel` to a `AbstractNLPModel` that can be solved with any standard NLP solver.
+
+# Arguments
+
+- `m::AbstractBlockNLPModel`: name of the BlockNLPModel
+"""
+function FullSpaceModel(::Type{T}, m::AbstractBlockNLPModel) where {T}
+    nb = m.problem_size.block_counter
+    n_var = sum(m.blocks[i].meta.nvar for i in 1:nb)
+
+    l_con::Vector{Float64} = []
+    u_con::Vector{Float64} = []
+    for i in 1:nb
+        if m.blocks[i].meta.ncon > 0
+            l_con = vcat(l_con, m.blocks[i].meta.lcon)
+            u_con = vcat(u_con, m.blocks[i].meta.ucon)
+        end
+    end
+    l_con = vcat(l_con, get_RHSvector(m))
+    u_con = vcat(u_con, get_RHSvector(m))
+
+    l_var::Vector{Float64} = []
+    u_var::Vector{Float64} = []
+    for i in 1:nb
+        l_var = vcat(l_var, m.blocks[i].meta.lvar)
+        u_var = vcat(u_var, m.blocks[i].meta.uvar)
+    end
+
+    meta = NLPModelMeta{T, Vector{T}}(
+        n_var,
+        ncon = sum(m.blocks[i].meta.ncon for i in 1:nb) + 
+        m.problem_size.link_counter,
+        nnzh = sum(m.blocks[i].meta.nnzh for i in 1:nb),
+        nnzj = sum(m.blocks[i].meta.nnzj for i in 1:nb) + 
+        nnz(get_blockmatrix(m)),
+        x0 = zeros(Float64, n_var),
+        lvar = l_var,
+        uvar = u_var,
+        lcon = l_con,
+        ucon = u_con,
+        minimize = true,
+        name = "full_space",
+    )
+return FullSpaceModel(meta, Counters(), m)
+end
+
+FullSpaceModel(m::AbstractBlockNLPModel) = FullSpaceModel(Float64, m)
+
+function NLPModels.obj(nlp::FullSpaceModel, x::AbstractVector)
+    nb = nlp.blocknlp.problem_size.block_counter
+    @lencheck nlp.meta.nvar x
+    increment!(nlp, :neval_obj)
+    return sum(obj(nlp.blocknlp.blocks[i].problem_block, 
+    x[nlp.blocknlp.blocks[i].var_idx]) for i in 1:nb)
+end
+
+function NLPModels.grad!(nlp::FullSpaceModel, x::AbstractVector, gx::AbstractVector)
+    nb = nlp.blocknlp.problem_size.block_counter
+    @lencheck nlp.meta.nvar x gx
+    increment!(nlp, :neval_grad)
+    for i in 1:nb
+        gx[nlp.blocknlp.blocks[i].var_idx] = 
+        grad(nlp.blocknlp.blocks[i].problem_block, x[nlp.blocknlp.blocks[i].var_idx])
```

Comment:
Can we modify `gx` inplace here by using `grad!`? In general, it's better to avoid any new allocation of a vector

Reply:
We can, I have updated the code. One question: Isn't `gx` already allocated here? and I am just assigning new values to its elements.

Reply:
I tried doing that, but it somehow doesn't work. Also, unless I am misunderstanding your comment, I don't think I am allocating any new memory here, because `gx` is already initialized as a `nlp.meta.nvar` sized vector.

Reply:
Okay, this is done now using `@view`.